contracts/oracle/PCVOracle.sol

```diff
@@ -0,0 +1,396 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.8.13;
+
+import {SafeCast} from "@openzeppelin/contracts/utils/math/SafeCast.sol";
+import {EnumerableSet} from "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import {Decimal} from "../external/Decimal.sol";
+
+import {IOracle} from "./IOracle.sol";
+import {VoltRoles} from "../core/VoltRoles.sol";
+import {CoreRefV2} from "../refs/CoreRefV2.sol";
+import {IPCVDepositV2} from "../pcv/IPCVDepositV2.sol";
+import {DynamicVoltSystemOracle} from "./DynamicVoltSystemOracle.sol";
+
+/// @notice Contract to centralize information about PCV in the Volt system.
+/// This contract will emit events relevant for building offchain dashboards
+/// of pcv growth, composition, and locations (venues).
+/// This contract keeps track of the percentage of illiquid investments,
+/// which allows dynamic VOLT rate based on the liquidity available for
+/// redemptions (in other parts of the system).
+/// Each PCV Deposit has an oracle, which allows governance to manually
+/// mark down a given PCVDeposit, if losses occur and the implementation
+/// does not automatically detect it & return erroneous balance() values.
+/// Oracles are also responsible for decimal normalization.
+/// @author Eswak, Elliot Friedman
+contract PCVOracle is CoreRefV2 {
+    using Decimal for Decimal.D256;
+    using SafeCast for *;
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    /// @notice emitted when a new venue oracle is set
+    event VenueOracleUpdated(
+        address indexed venue,
+        address indexed oldOracle,
+        address indexed newOracle
+    );
+
+    /// @notice emitted when a new venue is added
+    event VenueAdded(address indexed venue, bool isIliquid, uint256 timestamp);
+
+    /// @notice emitted when a venue is removed
+    event VenueRemoved(
+        address indexed venue,
+        bool isIliquid,
+        uint256 timestamp
+    );
+
+    /// @notice emitted when total venue PCV changes
+    event PCVUpdated(
+        address indexed venue,
+        bool isIliquid,
+        uint256 timestamp,
+        uint256 oldLiquidity,
+        uint256 newLiquidity
+    );
```

Comment:
This should be moved to the interface

Reply:
This is one in #162 ([7c8a09e](https://github.com/volt-protocol/volt-protocol-core/pull/162/commits/7c8a09e65c3c3b109f98d0eae24325654f7b3b76)), we should really merge everything together and work on new pull requests from there because it's a mess :-/